sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultInputFile.java

```diff
@@ -21,274 +21,240 @@
 
 import com.google.common.base.Preconditions;
 import java.io.File;
-import java.io.StringReader;
 import java.nio.charset.Charset;
 import java.nio.file.Path;
 import java.util.Arrays;
+import java.util.function.Function;
+
 import javax.annotation.CheckForNull;
-import javax.annotation.Nullable;
 import org.sonar.api.batch.fs.InputFile;
 import org.sonar.api.batch.fs.TextPointer;
 import org.sonar.api.batch.fs.TextRange;
-import org.sonar.api.batch.fs.internal.FileMetadata.Metadata;
-import org.sonar.api.utils.PathUtils;
 
 /**
  * @since 4.2
  */
 public class DefaultInputFile extends DefaultInputComponent implements InputFile {
-
-  private final String relativePath;
-  private final String moduleKey;
-  private Path moduleBaseDir;
-  private String language;
-  private Type type = Type.MAIN;
+  private final DefaultIndexedFile indexedFile;
+  private final Function<DefaultInputFile, Metadata> metadataGenerator;
   private Status status;
-  private int lines = -1;
   private Charset charset;
-  private int lastValidOffset = -1;
-  private String hash;
-  private int nonBlankLines;
-  private int[] originalLineOffsets;
+  private Metadata metadata;
 
-  public DefaultInputFile(String moduleKey, String relativePath) {
-    this.moduleKey = moduleKey;
-    this.relativePath = PathUtils.sanitize(relativePath);
+  public DefaultInputFile(DefaultIndexedFile indexedFile, Function<DefaultInputFile, Metadata> metadataGenerator) {
+    this.indexedFile = indexedFile;
+    this.metadataGenerator = metadataGenerator;
+    this.metadata = null;
+  }
+
+  private void checkMetadata() {
```

Comment:
Shouldn't we already prepare for multi-threading? This one would need to be synchronized.